### ISSUE TITLE
feature request - different client filter configuration

### ISSUE THREAD

**Issue by Kleinhirn** [AdguardTeam/AdGuardHome#435]
Hi,

is it possible to add a function for different client filter configuration.
I want to set different filters for my pc`s in network - do you can make this?
- childen pc with parent control
- different filter for PC/Laptop/Smartphone

Thanks

**Comment by AnthonyBe** [comment]
Was thinking same.

Specifically around children, being able to group their devices by MAC address and then provide the additional safe search and parental control settings on that group.

**Comment by alexsannikov** [comment]
That would be great to have specific settings for specific device based on devicename/MAC/IP.
I would apply Parent control & multiple filtering lists for kid's tablet, but allowed some resources for my smart or PC.

**Comment by Orava2** [comment]
I would like to set different upstream DNS servers for children and adults.

**Comment by ameshkov** [comment]
So here's what is requested:

1. General settings (browsing security / safe search / parental control) per client
2. Different filters per client
3. Different DNS servers per client

Number 1 is relatively easy. 2 and 3 are quite harder, though.

**Comment by ameshkov** [comment]
Number 1 is done: #727

**Comment by ArchiveBase** [comment]
https://github.com/AdguardTeam/AdGuardHome/issues/962

**Comment by vtolstov** [comment]
Any ETA for this issue ?

**Comment by ameshkov** [comment]
@vtolstov 1 and 3 are done.

What for 2, most likely we'll use a different approach to it -- see https://github.com/AdguardTeam/AdGuardHome/issues/1081

**Comment by ArchiveBase** [comment]
@ameshkov https://github.com/AdguardTeam/AdGuardHome/issues/1081 Isn't it annoying? i thought both features will be implemented.
I prefer suggest to implement this one similar google compute engine and firewall. https://github.com/AdguardTeam/AdGuardHome/issues/817#issuecomment-534275463 or any better ideas are welcome
- VM instances = Clients
- VPC network firewall rules = Filters

https://cloud.google.com/vpc/docs/using-firewalls
https://cloud.google.com/vpc/docs/add-remove-network-tags

**Comment by zejar** [comment]
Now that Pihole 5.0 has been released with group-based filters it would be great if AdGuard Home would also support setting filters based on clients/networks :)

**Comment by lordraiden** [comment]
@ameshkov 
Is there an ETA to have this available? or is not even a priority?
I see it has not milestone and is a quite useful feature.

**Comment by ameshkov** [comment]
I have not yet decided how we should approach this. In order to do this, I'd like to know how exactly you'd like to use this.

Here's one example, unblocking a couple of domains for a specific device: https://github.com/AdguardTeam/AdGuardHome/issues/1716

I need more examples to make a decision.

**Comment by AnthonyBe** [comment]
I had a specific use case just this week.
I work for a company whose primary domain name is blocked by AG's default filters. I've had to unblock a lot of subdomains just so I can work (or fire up VPN to bypass AGH).
I would've loved the ability to just nominate my laptop's IP address to bypass some filters, or white list just for my IP the root domain that was being blocked.
I know I can likely achieve same by creating a client group, then setting a ctag, and then manually creating some allow list entries with the ctag, but it was just too complicated for me to mess around with.
As it was, I just ended up turning AGH off for 5 mins to get something done and then turned it back on!

**Comment by ameshkov** [comment]
> I would've loved the ability to just nominate my laptop's IP address to bypass some filters, or white list just for my IP the root domain that was being blocked.

Yep, got it.

This makes sense, and I am thinking that the easiest way would be to simply allow adding rules for specific clients. Something like this: `@@||example.org^$client=192.168.1.1`

We can make it even easier by adding an option to do this right from the query log.

**Comment by Archive5** [comment]
> I have not yet decided how we should approach this. In order to do this, I'd like to know how exactly you'd like to use this.
> I need more examples to make a decision.

The problem with ctag; it's suitable for permanent rules something like printers etc as discussed in the feature request https://github.com/AdguardTeam/AdGuardHome/issues/1081  but if we maintain them frequently as add/remove rules from lists, assign for different devices time to time or when adult person and kids use same device at home then its too complicated with existing feature

So i suggest to add an option for write prefer tag names when create lists then we can assign them to devices through client settings page. that's how google cloud platform has designed to work with firewall rules

(Allow user to add multiple tag names to each list and use same tag name to multiple lists)

**Comment by lordraiden** [comment]
From a user experience view I think the best way would be

Every time you add a new block list you have the option to choose if:
1.1 This list will be apply by default to any new asset not tagged or defined as a client (Checkbox)  (if nothing is selected it will apply to everything, always like it is now)
1.2 Then deffined to what applies with this options
1.2.1 Define the clients to which the block list applies (clients defined already in AGH)
1.2.2 Define the tags to which the block list applies (tags for clients are defined in the client section)
1.2.3 Define IP ranges or subnets to which the block list applies (this will override 1.1)

The same logic will apply to "custom filtering rules" or "whitelisting" sections

**Comment by ameshkov** [comment]
@lordraiden it is not a question of how to add a client-specific list, I need to know when and why exactly you would use that (or something like that).

**Comment by lordraiden** [comment]
> 
> 
> @lordraiden it is not a question of how to add a client-specific list, I need to know when and why exactly you would use that (or something like that).

Ok,

Well I have dockers, vm and normal users at home.

In dockers and servers I want to use blocklists that are only related to malware domains. I don't care about phishing, or ads (depending on the product used) because no one is going to browse from those IP's. Even if it is a service that phone home like plex, I only need to block 1 domain for that specific service (analytics.plex.tv)

Another case would be IoT (TVs, Alexa, etc..) Maybe in this case I only want to block certain domains or use a specific list for my samsung TV

Kids maybe I can load to them a specific domain list to block p0rn xD (energized lists)

In more complex environments probably there are more use cases, and if in a future you want block website categories (I think AGH or the user would need to pay for those lists) you have half of the work done since you can define a blocking policy and apply that policy to clients.
Free and probably a poor source but maybe better than nothing  https://dsi.ut-capitole.fr/blacklists/index_en.php

Talking about policies, maybe this would be a cleaner way, you define a complete policy (blocklists, DNS, Whitelist, etc. with all the options and then you apply that policy to the clients you want.

**Comment by lordraiden** [comment]
BTW I have eddited my previous comment, take a quick look or maybe you read the latest
One more thing, another interesting use case would be to allow only a client to contact certain domains.
I think it would be quite useful for IoT devices, for example if I have a device that only calls to *abc.com, I would like to define that for that client only *abc.com is allowed so if the deviced gets compromised or does something weird it will be blocked.
In addition many server services, plex, duplicati, etc... can have a very limited range of domains that are in use so I can do a quick search in the logs over a month period, see all the domains they called and setup and allow list and block everyhing else. 

This will probably require a different approach since we are not working with blocklists and allow everyhing else but we define a whitelist and block everything else. But with the policy concept will be easy to define the logic (at least in the ui), it will be like a enterprise grade proxy/firewall, all of them use policies that are applied to something fw rules, users etc.

Maybe I'm a little bit security paranoid xD

**Comment by brendan-intermission** [comment]
> 
> 
> BTW I have eddited my previous comment, take a quick look or maybe you read the latest
> One more thing, another interesting use case would be to allow only a client to contact certain domains.
> I think it would be quite useful for IoT devices, for example if I have a device that only calls to *abc.com, I would like to define that for that client only *abc.com is allowed so if the deviced gets compromised or does something weird it will be blocked.
> In addition many server services, plex, duplicati, etc... can have a very limited range of domains that are in use so I can do a quick search in the logs over a month period, see all the domains they called and setup and allow list and block everyhing else.
> 
> This will probably require a different approach since we are not working with blocklists and allow everyhing else but we define a whitelist and block everything else. But with the policy concept will be easy to define the logic (at least in the ui), it will be like a enterprise grade proxy/firewall, all of them use policies that are applied to something fw rules, users etc.
> 
> Maybe I'm a little bit security paranoid xD

I second this, its very good security practice. These days we have such a random bunch of IOT/Smart home devices within the bounds of our internet network. It would be great to whitelist exactly which URLs/Call Home destinations they need. They should not need to connect to any other domains or be part of someones botnet to launch a DDoS.

**Comment by mHatsune** [comment]
https://github.com/AdguardTeam/AdGuardHome/issues/1880 https://github.com/AdguardTeam/AdGuardHome/issues/951

**Comment by mHatsune** [comment]
https://github.com/AdguardTeam/AdGuardHome/issues/1203

**Comment by lordraiden** [comment]
@ameshkov Will you asign a milestone to this?

**Comment by ameshkov** [comment]
Nope, I use this issue as a place for discussion and create new relevant issues for that.

Meanwhile, v0.103.3 adds `$client` modifier support: https://github.com/AdguardTeam/AdGuardHome/wiki/Hosts-Blocklists#-client

This lets you create client-specific filtering rules.

Next steps:

1. Allow creating client-specific rules from UI (block/unblock for this client)
2. Allow setting client-specific rules in the client settings

**Comment by lordraiden** [comment]
> Nope, I use this issue as a place for discussion and create new relevant issues for that.
> 
> Meanwhile, v0.103.3 adds `$client` modifier support: https://github.com/AdguardTeam/AdGuardHome/wiki/Hosts-Blocklists#-client
> 
> This lets you create client-specific filtering rules.
> 
> Next steps:
> 
>     1. Allow creating client-specific rules from UI (block/unblock for this client)
> 
>     2. Allow setting client-specific rules in the client settings

Any news on this?
It's in the roadmap?
It would be something like this  https://docs.pi-hole.net/database/gravity/example/

**Comment by ameshkov** [comment]
> 1. Allow creating client-specific rules from UI (block/unblock for this client)

Done in v0.104

**Comment by MZGSZM** [comment]
My use case for having per-device filters would be using more conservative lists for my family member's devices compared to my own. I currently use a regex list that catches more stuff on top of my main "ad and tracker blocker" list but it also has more false positives. I don't mind taking the time to whitelist a few things here and there as needed but it's a bit of a hassle when I'm not home and someone else comes across a false positive. So, I'd disable that list for select devices on the network.

**Comment by zzyonn** [comment]
So, for the moment there is no way to create group for user and specify list to this group ?

**Comment by Cybercave** [comment]
[Blocky](https://github.com/0xERR0R/blocky) has support for client blocklists. And it is in Go, too.

https://0xerr0r.github.io/blocky/configuration/#client-groups
![image](https://user-images.githubusercontent.com/14866477/134964462-74f56593-1696-4cbe-8100-689573569fd5.png)

**Comment by privacy-advo** [comment]
И на солнце есть пятна ...
[The new AdGuard DNS product](https://adguard-dns.com/en/welcome.html)
"_Different options for all purposes
You may want to have different configurations for your kids' tablets, your work laptop, and your smart TV. With AdGuard DNS you can have an individual setup for each of your devices._"

@ameshkov - Will we see this feature in AdGuard Home as well... ?

![grafik](https://user-images.githubusercontent.com/62058808/147495043-4998640e-6a06-4bb1-8db4-49a41d6246b3.png)

![grafik](https://user-images.githubusercontent.com/62058808/147495018-c4add8a0-7820-4256-ae36-a7ccbe347c42.png)

**Comment by ameshkov** [comment]
@privacy-advo we will definitely see it in AGH as well in the future. The new AG DNS was created from scratch so we could implement some of the features requested for AGH rather easily. In the case of AGH, the same requires some serious code rework. It is, of course, is going to happen, just trying to explain why it happen to AG DNS first.

**Comment by privacy-advo** [comment]
That's great news, can't wait to test the entirely new built system. I'm on the waiting list since the blog post. Actually it's very reasonable from a business perspective that you put your resources first into a new product. Nevertheless, I _hope_ that you're price model won't challenge me to stop running my own (cheap) VPS AdGuard Home instances and go entirely for new AG DNS. (Don't get me wrong, I own 25 lifetime licenses which cover devices of family and friends.)

**Comment by privacy-advo** [comment]
@ameshkov commented on 6 Apr 2019:
> So here's what is requested:
> 
>     1. General settings (browsing security / safe search / parental control) per client
> 
>     2. Different filters per client
> 
>     3. Different DNS servers per client
> 

> 
> Number 1 is relatively easy. 2 and 3 are quite harder, though.

Could _2. Different filter**lists** per client_ be an option for a milestone? It's still listed as P2.

**Comment by ameshkov** [comment]
> Could 2. Different filterlists per client be an option for a milestone? It's still listed as P2.

Yeah, we have this in our future plans.

**Comment by privacy-advo** [comment]
As 0.108 is on the horizon, is there a chance we will see P2 #435?

**Comment by ameshkov** [comment]
Yes, there is:)

**Comment by zzyonn** [comment]
Upgraded to 0.108 Beta, but not found this possibility :s

**Comment by ameshkov** [comment]
There is a chance, it's not done yet:)

**Comment by zzyonn** [comment]
> There is a chance, it's not done yet:)

Oupss sorry, my bad ..

**Comment by hoang-rio** [comment]
Waiting for this feature ❤️

**Comment by privacy-advo** [comment]
It's been almost 4 years. It would be great to finally see the feature.
v0.107.16? 👍

**Comment by useronegit** [comment]
I really hope it will be added soon, for me it would be the best feature for Adguard Home

**Comment by privacy-advo** [comment]
+1 
Is there a way the community could make it happen?

**Comment by gyCfjSnO** [comment]
> It's been almost 4 years. It would be great to finally see the feature. v0.107.16? 👍

Almost 5 years now 😂

**Comment by privacy-advo** [comment]
> > It's been almost 4 years. It would be great to finally see the feature. v0.107.16? 👍
> 
> Almost 5 years now 😂

Yeah. You name it. Pretty frustrating to see other issues being implemented with every release and this P2 issue, graved by so many as one can see all the mentions, lasts in the void ...

**Comment by Knot3n** [comment]
This is def. needed to have services control for children .. e.g TikTok .. Whatsapp for specific timeframes...

**Comment by gyCfjSnO** [comment]
> This is def. needed to have services control for children .. e.g TikTok .. Whatsapp for specific timeframes... @ameshkov could you give maybe an update about this ?

You can control TikTok, Instagram, WhatsApp to be blocked/allowed at specific times.

However, you still can't apply custom block lists to clients.

**Comment by AnthonyBe** [comment]
When I commented on this 5 years ago, it was primarily to better manage the devices of my teenage son.

That son is now an adult.

<sigh>

**Comment by Knot3n** [comment]
> > This is def. needed to have services control for children .. e.g TikTok .. Whatsapp for specific timeframes... @ameshkov could you give maybe an update about this ?
> 
> You can control TikTok, Instagram, WhatsApp to be blocked/allowed at specific times.
> 
> However, you still can't apply custom blocklists/filters to different clients. For example applying a blocklists/filter only to clients a, b, c while the rest of the clients aren't impacted.

Yeah sorry, i meant for a specific client not only a specific time.

**Comment by adrianmihalko** [comment]
C'mon developers, you can do it.

**Comment by bvandevliet** [comment]
I'd also like to have this feature! A newer issue (#4334) related to this was opened later on but got also closed already, don't know why..

**Comment by lux5am** [comment]
I want this too. I want to allow all ads on specific device but still enable a short list of custom filters for that device. Currently it's not possibly, there's only 1 toggle to enable/disable global filters and custom filters. I setup nextdns server just for this at the moment.

**Comment by RainmakerRaw** [comment]
Six years later... It's such a shame, because per-client filter list selection is the one killer feature AGH is missing. I ended up giving up, and now run Blocky on my LAN. It too is a single GO binary, one static YAML file, and I can list filters and assign clients or groups to them. Now ads and malware are blocked by default, but for kids also gambling, NSFW, enforced safesearch and piracy sites are blocked. I really hope you guys manage to finally integrate this some day, whether in the current AGH or the (at some point) AG DNS based rewrite.

**Comment by hoang-rio** [comment]
I don't think it's a techincal challenge. It's business problem

**Comment by privacy-advo** [comment]
> Six years later... It's the one killer feature AGH is missing. 

@ameshkov @ainar-g
Any chance we will see this? Maybe even a cancellation would help people. We _really_ are waiting for it.

**Comment by ainar-g** [comment]
It is still planned.  The main challenge is that the current filtering infrastructure is not designed to work with any custom engines.  The design for a new system is ongoing, but it's still a challenge, as there are a lot of things to take into account, including some of the features being in conflict with each other.  Simply copying filtering engines for each client would increase RAM consumption, for example, and we're already having issues reported about that every once in a while.  Additionally, there are requirements of thread safety and general performance, e.g. filters updating must not block DNS processing for all other clients or that two clients shouldn't end up with different snapshots of the same filter.

In the meantime, there are workarounds:
* The blocked services feature has had an upgrade, and in case some following this issue had missed that, you can assign blocked services per-client and also schedule them.  So blocking e.g. TikTok during a certain time of day has been possible for a few months now.
* The [`$client` filtering-rule syntax](https://adguard-dns.io/kb/general/dns-filtering-syntax/#client) also still works, and if you don't have that many rules, you can put them into a list.

**Comment by AnthonyBe** [comment]
Thanks @ainar-g , appreciate the explanation.

Equally, AG 'made' this problem by building some key features of AGH's functionality around the presumption of people using it's built-in DHCP server. Ironically, the built-in DHCP server cannot be used by users of AGH on Windows (feature not supported), so this isn't even a matter of choice as to _whether_ to use AGH DHCP (yes, I could choose to use a different OS, but that is now compounding a bigger problem/solution).

Your community of users (including myself here https://github.com/AdguardTeam/AdGuardHome/issues/961) flagged this back in 2019, as a design direction that would create a lot of limitations (to put it politely :) ). There was _ample_ time and opportunity to change course of AGH development. But you didn't. So, here we (still) are.

I am in no way implying that this is an easy problem to solve. I've been using AG products for a lot of years and I know you don't shy away from tough challenges. I imagine this is a conversation that more likely resolved around time, effort, priority, ROI, breadth of benefit, etc. I've worked in IT most of my life, so I get it. I am half-expecting the "we need to re-write the AGH code from scratch, to fix this, pull in a lot of the on-going development features from the public AG DNS product, and allow AGH and AG DNS to share common feature through more common code between the products".

**Comment by privacy-advo** [comment]
Thanks @ainar-g 
I'm very happy about `$client`. Not being able to use different configs made me write a 5k+ regex `$client` list myself over the years. : )
Nevertheless this eats up a lot of time, highly anticipate the feature. Good to hear it's still on the horizon.

(If AG paid DNS wouldn't have a 1k limit, I would use it. But when using default deny, 1k is not enough.)

**Comment by hoang-rio** [comment]
I'm try implementting this feature by my self at https://github.com/abpvn/AdGuardHome/releases
This implement just like run multiple AdGuardHome server by clone current filtering engine per client. It's may got performance issue (memory leak, high CPU ussage) when running with a lot of client.

Someone does not enough pattient to waiting for new optimized filtering engine for per client filter list can try my implement. It's not the best but it work

**Comment by ubergeek77** [comment]
Would it be reasonable in the current filtering system to, say:

- Spin up my own custom service that pulls an upstream blocklist and adds a `$client` of my choice to each entry
- Put the URL of this custom service in AGH as a blocklist, instead of the "real" blocklist

Is that the best approach for now, or is `$client` an expensive operation that should only be used sparingly?

Happy to hear any clever workarounds people have thought of as well.

**Comment by privacy-advo** [comment]
> Six years later... the one killer feature AGH is missing.

@ameshkov @ainar-g
Any chance we will see this 2025?

**Comment by WolframCrem** [comment]
any chance of seeing per client allow / blocklists?

**Comment by privacy-advo** [comment]
@ameshkov @ainar-g 
Is that the subject of the table?

**Comment by ainar-g** [comment]
It is still planned.  No exact timeframes, sorry.  `$client` is currently the recommended way to customize your filters depending on the client.

**Comment by privacy-advo** [comment]
Great, thanks for the followup. The userbase values this feature request as there are persistent messages about it.

**Comment by fegauthier-paragon** [comment]
I did a PR for per-client blocklist / allowlist

https://github.com/AdguardTeam/AdGuardHome/pull/8274

**Comment by privacy-advo** [comment]
@ainar-g, could this be the end of an eight-year history of longing and unpleasant workarounds?

**Comment by privacy-advo** [comment]
@ameshkov Do you have news on this? It's been 8 years soon.
